modin/logging/config.py

```diff
@@ -0,0 +1,127 @@
+import logging
+import datetime as dt
+import os
+import uuid
+import platform
+import psutil
+import pkg_resources
+import threading
+import time
+from modin.config import LogMode
+
+__LOGGER_CONFIGURED__: bool = False
+
+
+class ModinFormatter(logging.Formatter):
+    def formatTime(self, record, datefmt=None):
+        """
+        Return the creation time of the specified LogRecord as formatted text.
+        This custom logging formatter inherits from the logging module and
+        records timestamps at the microsecond level of granularity.
+
+        Parameters
+        ----------
+        record: LogRecord
+            The specified LogRecord object.
+        datefmt: str, default: None
+            Used with time.ststrftime() to format time record.
+
+        Returns
+        -------
+        datetime
+            datetime object containing microsecond timestamp.
+        """
+        ct = dt.datetime.fromtimestamp(record.created)
+        if datefmt:
+            s = ct.strftime(datefmt)
+        else:
+            # Format datetime object ct to microseconds
+            t = ct.strftime("%Y-%m-%d %H:%M:%S")
+            s = f"{t},{record.msecs:03}"
+        return s
+
+
+def bytes_int_to_str(num_bytes, suffix="B"):
+    """
+    Scale bytes to its human-readable format (e.g: 1253656678 => '1.17GiB').
+
+    Parameters
+    ----------
+    num_bytes: int
+        Number of bytes.
+    suffix: str, default: "B"
+        Suffix to add to conversion of num_bytes.
+
+    Returns
+    -------
+    str
+        Human-readable string format.
+    """
+    factor = 1024
+    if num_bytes > 1000000000000000000:
+        raise ValueError("System memory exceeds expectations")
+    for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi"]:
+        if num_bytes < factor:
+            return f"{num_bytes:.2f}{unit}{suffix}"
+        num_bytes /= factor
+
+
+def configure_logging(level):
+    global __LOGGER_CONFIGURED__
+    logger = logging.getLogger("modin.logger")
+    job_id = uuid.uuid4().hex
+    log_filename = f".modin/logs/job_{job_id}.log"
+
+    if not os.path.isdir(".modin/logs"):
+        os.makedirs(os.path.dirname(log_filename), exist_ok=False)
+
+    logfile = logging.FileHandler(log_filename, "a")
+    formatter = ModinFormatter(
+        fmt="%(process)d, %(thread)d, %(asctime)s, %(message)s",
+        datefmt="%Y-%m-%d,%H:%M:%S.%f",
+    )
+    logfile.setFormatter(formatter)
+    logger.addHandler(logfile)
+    logger.setLevel(level)
+
+    __LOGGER_CONFIGURED__ = True
+
+
+def memory_thread(logger):
```

Comment:
question: given we may want to log the memory usage of distributed execution, how would it change the current design? would we have to use something like psutil processes?

Reply:
I don't think `psutil` would help in this case, but we have access to the Ray Object IDs, so we could asynchronously hit the Ray API and access some of the object's attributes. The challenge here is the Ray API may take some time to return so we should set a timeout or have this running in another separate thread that ModinVision parses out later from the logs.